src/repositories/docsetsRepository.ts

```diff
@@ -0,0 +1,119 @@
+import { Db } from 'mongodb';
+import { BaseRepository } from './baseRepository';
+import { ILogger } from '../services/logger';
+import { IConfig } from 'config';
+
+const docsetsCollectionName = process.env.DOCSETS_COL_NAME || 'docsets';
+export class DocsetsRepository extends BaseRepository {
+  constructor(db: Db, config: IConfig, logger: ILogger) {
+    super(config, logger, 'DocsetsRepository', db.collection(docsetsCollectionName));
+  }
+
+  private getAggregationPipeline(
+    matchConditionField: string,
+    matchConditionValue: string,
+    projection?: { [k: string]: number }
+  ) {
+    return [
+      // Stage 1: Unwind the repos array to create multiple documents for each referenced repo
+      {
+        $unwind: '$repos',
+      },
+      // Stage 2: Lookup to join with the repos_branches collection
+      {
+        $lookup: {
+          from: 'repos_branches',
+          localField: 'repos',
+          foreignField: '_id',
+          as: 'repo',
+        },
+      },
+      // Stage 3: Match documents based on given field
+      {
+        $match: {
+          [`repo.${matchConditionField}`]: matchConditionValue,
+        },
+      },
+      // Stage 4: Merge/flatten repo into docset
+      {
+        $replaceRoot: { newRoot: { $mergeObjects: [{ $arrayElemAt: ['$repo', 0] }, '$$ROOT'] } },
+      },
+      // Stage 5: Exclude fields
+      {
+        $project: projection || {
+          _id: 0,
+          repos: 0,
+          repo: 0,
+        },
+      },
+    ];
+  }
+
+  async getProjectByRepoName(repoName: string) {
+    const projection = { project: 1 };
+    const aggregationPipeline = this.getAggregationPipeline('repoName', repoName, projection);
+    const cursor = await this.aggregate(aggregationPipeline, `Error while getting project by repo name ${repoName}`);
+    const res = await cursor.toArray();
+    if (!res.length) {
+      const msg = `DocsetsRepository.getProjectByRepoName - Could not find project by repoName: ${repoName}`;
+      this._logger.info(this._repoName, msg);
+    }
+    return res[0].project;
+  }
+
+  async getRepo(repoName: string): Promise<any> {
+    const aggregationPipeline = this.getAggregationPipeline('repoName', repoName);
```

Comment:
as you mentioned in [this comment](https://github.com/mongodb/docs-worker-pool/pull/909/files#r1331827811), are we still going to query `docsets` for getting repo info? 

`getProjectByRepoName` seems like it should be in `docsets` as `project` field should only be there in the future (CMIIW)

but `getRepo` and `getRepoBranchesByRepoName` seem like they can refer to the `REPOS_BRANCHES` collection

Reply:
These queries are being used to access properties that are only available in the docsets collection as well as properties that are in the repos_branches collection! They need properties such as `bucket` and `url`

Reply:
gotcha! so these are returning (essentially what is currently) a repo_branches entry, but it will be constructed from `docsets` since we are aiming to deprecate them from repos branches 👍

Reply:
Exactly!